File: Journal/backend/app/schemas/reflection.py
"""
Reflection API request/response schemas.

Defines Pydantic schemas for validating incoming requests
and serializing outgoing responses for reflection endpoints.
"""

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel

from app.models.reflection import RangeType


class ReflectionGenerateRequest(BaseModel):
    """Schema for requesting a new AI-generated reflection.

    Attributes:
        range_type: The date range type to analyse.
        start_date: Custom range start (required when range_type is 'custom').
        end_date: Custom range end (required when range_type is 'custom').
    """

    range_type: RangeType
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None


class ReflectionResponse(BaseModel):
    """Schema for a single reflection / emotional summary response.

    Attributes:
        id: The summary's unique identifier.
        summary: The AI-generated overall summary.
        emotional_patterns: Detected emotional patterns.
        positive_observations: Positive observations noted.
        gentle_insights: Gentle insights for the user.
        growth_suggestions: Growth suggestions.
        selected_range: The date range type used.
        created_at: Creation timestamp.
    """

    id: str
    summary: str
    emotional_patterns: List[str]
    positive_observations: List[str]
    gentle_insights: List[str]
    growth_suggestions: List[str]
    selected_range: str
    created_at: datetime


class ReflectionListResponse(BaseModel):
    """Schema for a list of reflections.

    Attributes:
        reflections: The list of reflection responses.
        total: Total number of reflections.
    """

    reflections: List[ReflectionResponse]
    total: int

File: Journal/frontend/src/types/reflection.ts
/* ──────────────────────────────────────────────────────────────
   Reflection / AI Summary Types
   ────────────────────────────────────────────────────────────── */

/** Predefined date range options for generating reflections */
export type RangeType =
  | "last_3_days"
  | "last_5_days"
  | "last_7_days"
  | "last_30_days"
  | "custom";

/** Human-readable labels for range types */
export const RANGE_LABELS: Record<RangeType, string> = {
  last_3_days:  "Last 3 Days",
  last_5_days:  "Last 5 Days",
  last_7_days:  "Last 7 Days",
  last_30_days: "Last 30 Days",
  custom:       "Custom Range",
};

/** All range types as an array */
export const ALL_RANGES: RangeType[] = [
  "last_3_days", "last_5_days", "last_7_days", "last_30_days", "custom",
];

/** Request body for generating an AI reflection */
export interface ReflectionGenerateRequest {
  range_type: RangeType;
  start_date?: string;
  end_date?: string;
}

/** A single AI-generated emotional reflection */
export interface EmotionalSummary {
  id: string;
  summary: string;
  emotional_patterns: string[];
  positive_observations: string[];
  gentle_insights: string[];
  growth_suggestions: string[];
  selected_range: {
    range_type: RangeType;
    start_date: string;
    end_date: string;
  };
  created_at: string;
}

/** Paginated list of reflections */
export interface ReflectionListResponse {
  reflections: EmotionalSummary[];
  total: number;
}
